expressions: Update/simplify controlling expressions (#1)
gklimowicz/fpp-202y: drafts/25-xxx-specifications.txt

5 Expressions allowed in #if and #elif directives
=================================================

ex05. Controlling expressions are made up of WHOLE_NUMBERs and

>       operators applied to subexpressions.

Reviewer: This rule now duplicates ex15, and by appearing first it could be misread to imply something about the pre-expansion token list. So I think it should just be removed.

```suggestion
```

ex10. When evaluating a #if or #elif directive, preprocessing
      evaluates and expands all object-like macros and function-like
      macro invocations to create a token-list of the expression to be

>       evaluated.

Reviewer: There's a critical missing step here, to evaluate the `defined` operator. This must take place BEFORE macro expansion.

I suggest we describe this as a multi-step process, eg:

```suggestion
ex05. When evaluating a #if or #elif directive, first every instance of the
      `defined ID` or `defined(ID)` operator is evaluated and replaced with 
      the resulting WHOLE_NUMBER. A `defined` expression evaluates to 1 if 
      the identifier is currently defined as a macro name (that is, if it is 
      predefined or if it has been the subject of a `#define` preprocessing 
      directive without an intervening `#undef` directive with the same 
      subject identifier), 0 if it is not.
      
ex10. When evaluating a #if or #elif directive and after `define` processing 
      described in ex05, the token-list is then subjected to macro expansion
      and replacement (see the section "Macro recognition and expansion").
      This results in a token-list of the expression to be
      evaluated. 
```

ex12. Since expression evaluation occurs *after* macro expansion, there will
      be no object-like macro or function-like macro invocations left to expand.
      All remaining IDs other than `true` (including those lexically identical
      to keywords such as `false`) are replaced with the WHOLE_NUMBER 0, true is
      replaced with the WHOLE_NUMBER 1, and then each preprocessing token is

>       converted into a token.

Reviewer: This sentence was copied from C23 CPP, but it doesn't work for us directly.

The requirements paper did not propose adding `true` and `false` to the FPP expression grammar. Those boolean literals were added in C23, but I'm pretty sure we resolved to leave that out (see `dfc84`).

Also, we've so far not defined a distinction between preprocessing tokens and other tokens, so that C23 terminology does not apply (at least for this revision of the document).

```suggestion
      All remaining IDs are replaced with the WHOLE_NUMBER 0.
```

      how the operators combine with sub-expressions containing other operators
      at different precedence levels.

op05. An "associativity" is assigned to each operator that determines how
      operators at the same precedence level are combined.
          - "left" means the operator binds to the left,
          - "right" means the operator binds to the right.
          - "nonassoc" means that the operator is not associative.

op07. The following table describes the semantics of the allowed operators
      in conditional expressions. The table is grouped by precedence level,
      from lowest precedence to highest.

      We label subexpressions "e1", "e2", and "e3" to aid in describing the
      evaluation semantics. Unless otherwise specified, all operators
      evaluate with the same semantics as their Fortran counterparts.

| Prec |   Op    |    Syntax    |  Assoc'y |  Evaluation Semantics      |
|------+---------+--------------+----------+----------------------------|
| low  |   ? :   | e1 ? e2 : e3 |  right   |      conditional-expr      |
|------+---------+--------------+----------+----------------------------|

> |      |   ¦¦    |   e1 || e2   |   left   |        logical OR          |

Reviewer: ```suggestion
|      |   ¦¦    |   e1 || e2   |   left   |    logical OR (see op12)   |
```

|------+---------+--------------+----------+----------------------------|

> |      |   &&    |   e1 && e2   |   left   |        logical AND         |

Reviewer: ```suggestion
|      |   &&    |   e1 && e2   |   left   |   logical AND (see op13)   |
```

|      | unary - |     - e1     |  right   |        unary -             |
|      | unary ~ |     ~ e1     |  right   |     Fortran NOT(e1)        |
|      | unary ! |     ! e1     |  right   | 1 if e1 == 0, 0 otherwise  |
|------+---------+--------------+----------+----------------------------|
| high | ( e1 )  |              |   N/A    |            e1              |
|------+---------+--------------+----------+----------------------------|

op12. The logical OR operator guarantees left-to-right evaluation; if the first

>       operand compares equal to 0, the second operand is not evaluated.

Reviewer: I think we need to fully define "logical OR" and "logical AND":

```suggestion
      operand evaluates to 1, the second operand is not evaluated and the
      resulting value is 1. Otherwise, the second operand is evaluated
      and provides the resulting value.
```

op13. The logical AND operator guarantees left-to-right evaluation; if the first

>       operand compares equal to 1, the second operand is not evaluated.

Reviewer: ```suggestion
      operand evaluates to 0, the second operand is not evaluated and the 
      resulting value is 0. Otherwise, the second operand is evaluated
      and provides the resulting value.      
```

|      | unary - |     - e1     |  right   |        unary -             |
|      | unary ~ |     ~ e1     |  right   |     Fortran NOT(e1)        |
|      | unary ! |     ! e1     |  right   | 1 if e1 == 0, 0 otherwise  |
|------+---------+--------------+----------+----------------------------|
| high | ( e1 )  |              |   N/A    |            e1              |
|------+---------+--------------+----------+----------------------------|

op12. The logical OR operator guarantees left-to-right evaluation; if the first

>       operand evaluates to 1, the second operand is not evaluated and the

Reviewer: ```suggestion
      operand evaluates to non-zero, the second operand is not evaluated and the
```
